src/go/debug-collector/dumper/dumper.go

```diff
@@ -0,0 +1,222 @@
+			return errors.Wrap(err, "get namespaces")
+		}
+
+		err = json.Unmarshal(output, &nss)
+		if err != nil {
+			return errors.Wrap(err, "unmarshal namespaces")
+		}
+	}
+
+	for _, ns := range nss.Items {
+		output, err := d.runCmd("get", "pods", "-o", "json", "--namespace", ns.Name)
+		if err != nil {
+			continue // runCmd already stored this error in Dumper.Errors
+		}
+		var pods k8sPods
+		err = json.Unmarshal(output, &pods)
+		if err != nil {
+			d.saveCommandError(err.Error(), "unmarshal pods from namespace", ns.Name)
+			log.Println(errors.Wrap(err, "unmarshal pods"))
+		}
+
+		for _, pod := range pods.Items {
+			location := d.location + "/" + ns.Name + "/" + pod.Name + "/logs.txt"
+			output, err = d.runCmd("logs", pod.Name, "--namespace", ns.Name, "--all-containers")
+			if err != nil {
+				err = createArchive(location, d.mode, []byte(err.Error()), tw)
+				if err != nil {
+					log.Printf("create archive with err: %v", err)
+				}
+				continue
+			}
+			err = createArchive(location, d.mode, output, tw)
+			if err != nil {
+				log.Printf("create archive for pod %s: %v", pod.Name, err)
+			}
+		}
+
+		for _, resource := range d.resources {
+			err = d.getResource(resource, ns.Name, tw)
+			if err != nil {
+				log.Println(errors.Wrapf(err, "get %s resource", resource))
+			}
+		}
+
+		err = d.writeErrorsToFile(tw)
+		if err != nil {
+			log.Println(errors.Wrap(err, "write errors"))
+		}
+	}
+
+	err = d.getResource("nodes", "", tw)
+	if err != nil {
+		log.Println(errors.Wrapf(err, "get nodes"))
+	}
+
+	return nil
+}
+
+// runCmd run command (Dumper.cmd) with given args, return it output and save all errors in Dumper.errors
+func (d *Dumper) runCmd(args ...string) ([]byte, error) {
+	var outb, errb bytes.Buffer
+	cmd := exec.Command(d.cmd, args...)
+	cmd.Stdout = &outb
+	cmd.Stderr = &errb
+	err := cmd.Run()
+	if err != nil {
+		d.saveCommandError(err.Error()+" "+outb.String(), args...)
+		return outb.Bytes(), err
+	}
+	if errb.Len() > 0 {
+		d.saveCommandError(errb.String()+" "+outb.String(), args...)
+		return outb.Bytes(), err
+	}
+
+	return outb.Bytes(), nil
+}
+
+func (d *Dumper) getResource(name, namespace string, tw *tar.Writer) error {
+	location := d.location
+	args := []string{"get", name, "-o", "yaml"}
+	if len(namespace) > 0 {
+		args = append(args, "--namespace", namespace)
+		location = d.location + "/" + namespace
+	}
+	location += "/" + name + ".yaml"
+	output, err := d.runCmd(args...)
+	if err != nil {
+		return createArchive(location, d.mode, []byte(err.Error()), tw)
+	}
+
+	return createArchive(location, d.mode, output, tw)
+}
+
+func (d *Dumper) saveCommandError(err string, args ...string) {
+	command := d.cmd + " " + strings.Join(args, " ")
+
+	d.Errors[command] = err
+}
+
+func (d *Dumper) writeErrorsToFile(tw *tar.Writer) error {
+	var errStr string
+	for cmd, errS := range d.Errors {
+		errStr += cmd + ": " + errS + "\n"
+	}
+
+	return createArchive(d.location+"/errors.txt", d.mode, []byte(errStr), tw)
+}
```

Comment:
You don't need to store errors into the map only to convert it into the massive string which then converted to the bytes slice. Create bytes.Buffer and write all errors directly there. And at the end of the day just send this buffer to the archive.

Reply:
https://github.com/percona/percona-toolkit/pull/452/commits/118110a671ae050d9ba09342b631a0fecf9a964e